Mathlib/LinearAlgebra/Matrix/SVD/svdReindex.lean

```diff
@@ -0,0 +1,120 @@
+
+/-- For matrix of size m × n nad rank A.rank : we have an equivalence relation between the elements
+of Fin (height) and the eigevalues of the matrix A⬝Aᴴ partitioned into
+- non-zero eigenvaules: (exactly A.rank) of them see `fin_rank_equiv_eigs_mul_conjTranspose`
+- zero eigenvaules: (exactly height - A.rank) of them -/
+noncomputable def equiv_fin_height_eigs_mul_conjTranspose (A: Matrix (Fin M) (Fin N) 𝕂) :
+    (Fin M) ≃ (Fin A.rank) ⊕ (Fin (M - A.rank)) := by
+  let em := Equiv.sumCompl (fun i =>  (isHermitian_mul_conjTranspose_self A).eigenvalues i ≠ 0)
+  let eᵣ' : {i // (isHermitian_mul_conjTranspose_self A).eigenvalues i ≠ 0} ≃ Fin A.rank := by
+    apply Fintype.equivFinOfCardEq
+    rw [rank_eq_card_pos_eigs_self_mul_conj_transpose]
+  let eₘᵣ : {i // ¬(isHermitian_mul_conjTranspose_self A).eigenvalues i ≠ 0} ≃
+    Fin (M - A.rank) := by
+    apply Fintype.equivFinOfCardEq
+    rw [Fintype.card_subtype_compl, Fintype.card_fin, rank_eq_card_pos_eigs_self_mul_conj_transpose]
+  exact Equiv.trans em.symm  (Equiv.sumCongr eᵣ' eₘᵣ)
+
+/-- When the eigenvalues of the matrix Aᴴ⬝A are partitioned using
+`equiv_fin_width_eigs_conjTranspose_mul_self` i.e. into non-zero and zero eigenvalues, any element
+from the second partition is obivuosly zero! -/
+lemma enz_nr_zero (A: Matrix (Fin M) (Fin N) 𝕂) (i: Fin (N - A.rank)):
+  (isHermitian_transpose_mul_self A).eigenvalues
+    ((equiv_fin_width_eigs_conjTranspose_mul_self A).symm (Sum.inr i)) = 0 := by
+  unfold equiv_fin_width_eigs_conjTranspose_mul_self Equiv.sumCongr
+  simp only [ne_eq, Equiv.symm_trans_apply, Equiv.symm_symm, Equiv.coe_fn_symm_mk, Sum.elim_inr,
+    Equiv.sumCompl_apply_inr]
+  let eₙᵣ : {i // ¬(isHermitian_transpose_mul_self A).eigenvalues i ≠ 0} ≃ Fin (N - A.rank) := by
+    apply Fintype.equivFinOfCardEq
+    rw [Fintype.card_subtype_compl, Fintype.card_fin, rank_eq_card_pos_eigs_conj_transpose_mul_self]
+  exact Iff.mp Function.nmem_support ((eₙᵣ.symm i).prop)
+
+/-- When the eigenvalues of the matrix A⬝Aᴴ are partitioned using
+`equiv_fin_height_eigs_mul_conjTranspose` i.e. into non-zero and zero eigenvalues, any element from
+the second partition is obivuosly zero! -/
+lemma emz_mr_zero (A: Matrix (Fin M) (Fin N) 𝕂) (i: Fin (M - A.rank)) :
+    (isHermitian_mul_conjTranspose_self A).eigenvalues
+      ((equiv_fin_height_eigs_mul_conjTranspose A).symm (Sum.inr i)) = 0 := by
+  unfold equiv_fin_height_eigs_mul_conjTranspose Equiv.sumCongr
+  simp only [ne_eq, Equiv.symm_trans_apply, Equiv.symm_symm, Equiv.coe_fn_symm_mk, Sum.elim_inr,
+    Equiv.sumCompl_apply_inr]
+  let eₘᵣ : {i // ¬(isHermitian_mul_conjTranspose_self A).eigenvalues i ≠ 0} ≃
+    Fin (M - A.rank) := by
+    apply Fintype.equivFinOfCardEq
+    rw [Fintype.card_subtype_compl, Fintype.card_fin, rank_eq_card_pos_eigs_self_mul_conj_transpose]
+  exact Iff.mp Function.nmem_support ((eₘᵣ.symm i).prop)
+
+lemma enz_inj (A: Matrix (Fin M) (Fin N) 𝕂) (i j: Fin (N - A.rank)) :
```

Comment:
I must say I don't see why we need `enz_inj` and `emz_inj`: injectivity of equivs and of `Sum.inr` is already known, so I'd prefer to use known injectivity results twice over this very specific application.